File: worker/worker/risk.py
"""정품 리스크 점수 — 확정 판정이 아니라 '의심 신호의 합'. 순수 함수만, 웹 lib/risk.ts 의 assessRisk 와 규칙을 같게 유지.

점수 → 등급: 60↑ high / 30↑ medium / 그 외 low. 사용자가 '정품 확인됨'(verified) 으로 표시하면 항상 low.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Optional, Sequence

# 강한 키워드: 상품 자체가 정품이 아님을 사실상 인정
STRONG_WORDS = ["레플", "레플리카", "replica", "미러급", "s급", "sa급", "1:1", "고퀄", "최고급퀄", "이미테이션", "짝퉁", "가품", "custom made", "리얼리티"]
# 약한 키워드: 정품이어도 붙을 수 있지만 브랜드 신품이 아님
WEAK_WORDS = ["호환", "compatible", "벌크", "bulk", "리퍼", "refurb", "병행", "병행수입", "중고", "used", "비품", "비정품", "unbranded", "generic", "oem"]
OFFICIAL_WORDS = ["공식", "브랜드스토어", "브랜드 스토어", "로켓", "직영", "official", "flagship", "authorized", "정품판매"]
OFFICIAL_SITES = {"bestbuy", "walmart", "target"}                 # 직영 리테일러
OVERSEAS_MARKET_SITES = {"aliexpress", "temu", "shein", "taobao", "tmall", "1688", "jd"}

PRICE_SELF_HIGH = 50.0    # 자체 기준선 대비 -50% 이상
PRICE_SELF_MID = 35.0
PRICE_CROSS_HIGH = 40.0   # 동일 모델 다른 상품 중앙값 대비 -40% 이상
PRICE_CROSS_MID = 25.0
SELLER_CHURN_HIGH = 5     # 90일 내 판매자 변경 5회↑ → +25
SELLER_CHURN_MID = 3      # 3회↑ → +15


@dataclass
class RiskResult:
    level: str                  # low | medium | high
    score: int
    reasons: list[str] = field(default_factory=list)

    @property
    def text(self) -> str:
        return " · ".join(self.reasons)


def _has(text: str, words: Sequence[str]) -> Optional[str]:
    t = text.lower()
    for w in words:
        if w.lower() in t:
            return w
    return None


def is_official_seller(seller: Optional[str], site: str) -> bool:
    if site in OFFICIAL_SITES:
        return True
    return bool(seller and _has(seller, OFFICIAL_WORDS))


def assess_risk(
    title: str,
    site: str,
    price: float,
    seller: Optional[str] = None,
    baseline: Optional[float] = None,
    cross_baseline: Optional[float] = None,   # 같은 model_no 를 가진 '다른' 상품들의 최신가 중앙값 (같은 통화)
    model_no: Optional[str] = None,
    usual_seller: Optional[str] = None,       # 이력상 최빈 판매자
    verified: bool = False,
    seller_changes: int = 0,                  # 최근 90일 판매자 변경 횟수 (baseline.seller_changes)
) -> RiskResult:
    if verified:
        return RiskResult("low", 0, ["사용자가 정품 확인함"])

    score = 0
    reasons: list[str] = []

    # 1) 제목 키워드
    w = _has(title, STRONG_WORDS)
    if w:
        score += 60; reasons.append(f"제목에 '{w}'")
    else:
        w = _has(title, WEAK_WORDS)
        if w:
            score += 15; reasons.append(f"제목에 '{w}'")

    # 2) 가격 이상치 — 자체 기준선
    if baseline and baseline > 0:
        pct = (baseline - price) / baseline * 100
        if pct >= 65:
            score += 60; reasons.append(f"평소 가격 대비 -{pct:.0f}%")
        elif pct >= PRICE_SELF_HIGH:
            score += 40; reasons.append(f"평소 가격 대비 -{pct:.0f}%")
        elif pct >= PRICE_SELF_MID:
            score += 20; reasons.append(f"평소 가격 대비 -{pct:.0f}%")

    # 3) 가격 이상치 — 동일 모델 교차 비교
    if cross_baseline and cross_baseline > 0:
        pct = (cross_baseline - price) / cross_baseline * 100
        if pct >= 60:
            score += 60; reasons.append(f"같은 모델 다른 판매처 대비 -{pct:.0f}%")
        elif pct >= PRICE_CROSS_HIGH:
            score += 40; reasons.append(f"같은 모델 다른 판매처 대비 -{pct:.0f}%")
        elif pct >= PRICE_CROSS_MID:
            score += 20; reasons.append(f"같은 모델 다른 판매처 대비 -{pct:.0f}%")

    # 4) 판매자
    official = is_official_seller(seller, site)
    if official:
        score -= 20
    elif usual_seller and is_official_seller(usual_seller, site) and seller and seller != usual_seller:
        score += 30; reasons.append(f"공식 판매자({usual_seller})에서 제3자({seller})로 바뀜")
    elif seller and seller not in ("직접 기록", "demo", "mock-seller"):
        score += 10; reasons.append("공식 판매자 아님")

    # 4b) 판매자 이력 — 자주 바뀌면 최저가 경쟁 셀러가 돌아가며 붙는 상품 (공식 판매자면 무시)
    if not official and seller_changes >= SELLER_CHURN_HIGH:
        score += 25; reasons.append(f"판매자가 자주 바뀜 (90일 {seller_changes}회)")
    elif not official and seller_changes >= SELLER_CHURN_MID:
        score += 15; reasons.append(f"판매자가 자주 바뀜 (90일 {seller_changes}회)")

    # 5) 해외 오픈마켓
    if site in OVERSEAS_MARKET_SITES:
        score += 15; reasons.append("해외 오픈마켓 (판매자 검증 약함)")

    # 6) 모델명 불일치
    if model_no and title and model_no.lower().replace("-", "") not in title.lower().replace("-", "").replace(" ", ""):
        score += 10; reasons.append(f"제목에 모델명 {model_no} 없음")

    score = max(0, score)
    level = "high" if score >= 60 else "medium" if score >= 30 else "low"
    if not reasons and level == "low":
        reasons.append("특이 신호 없음")
    return RiskResult(level, score, reasons)


MODEL_TOKEN_RE = re.compile(r"\b[A-Z]{1,4}-?\d{2,5}[A-Z]{0,3}\d{0,2}\b")


def guess_model_no(title: str) -> Optional[str]:
    """제목에서 모델명처럼 보이는 토큰 하나 (예: WH-1000XM6, OLED65C5, MX Master → 없음)."""
    m = MODEL_TOKEN_RE.search(title.upper())
    return m.group(0) if m else None

File: web/src/lib/risk.ts
/** 정품 리스크 — 워커 risk.py 와 같은 규칙. 확정 판정이 아니라 '의심 신호의 합'. */
import type { Snapshot } from "../types";
export type RiskLevel = "low" | "medium" | "high";
export interface RiskResult { level: RiskLevel; score: number; reasons: string[] }

const STRONG = ["레플", "레플리카", "replica", "미러급", "s급", "sa급", "1:1", "고퀄", "최고급퀄", "이미테이션", "짝퉁", "가품", "custom made", "리얼리티"];
const WEAK = ["호환", "compatible", "벌크", "bulk", "리퍼", "refurb", "병행", "병행수입", "중고", "used", "비품", "비정품", "unbranded", "generic", "oem"];
const OFFICIAL = ["공식", "브랜드스토어", "브랜드 스토어", "로켓", "직영", "official", "flagship", "authorized", "정품판매"];
const OFFICIAL_SITES = new Set(["bestbuy", "walmart", "target"]);
const OVERSEAS = new Set(["aliexpress", "temu", "shein", "taobao", "tmall", "1688", "jd"]);
const NEUTRAL_SELLERS = new Set(["직접 기록", "demo", "mock-seller"]);
const SELLER_CHURN_HIGH = 5, SELLER_CHURN_MID = 3;   // 90일 내 판매자 변경 횟수 → +25 / +15

const has = (text: string, words: string[]) => { const t = text.toLowerCase(); return words.find((w) => t.includes(w.toLowerCase())) ?? null; };
export const isOfficialSeller = (seller: string | null | undefined, site: string) => OFFICIAL_SITES.has(site) || !!(seller && has(seller, OFFICIAL));

export function assessRisk(o: {
  title: string; site: string; price: number | null; seller?: string | null; baseline?: number | null;
  crossBaseline?: number | null; modelNo?: string | null; usualSeller?: string | null; verified?: boolean;
  sellerChanges?: number;   // countSellerChanges() — 최근 window 내 판매자 변경 횟수
}): RiskResult {
  if (o.verified) return { level: "low", score: 0, reasons: ["사용자가 정품 확인함"] };
  let score = 0; const reasons: string[] = [];
  const sw = has(o.title, STRONG);
  if (sw) { score += 60; reasons.push(`제목에 '${sw}'`); }
  else { const ww = has(o.title, WEAK); if (ww) { score += 15; reasons.push(`제목에 '${ww}'`); } }

  if (o.price != null && o.baseline && o.baseline > 0) {
    const pct = ((o.baseline - o.price) / o.baseline) * 100;
    if (pct >= 65) { score += 60; reasons.push(`평소 가격 대비 -${pct.toFixed(0)}%`); }
    else if (pct >= 50) { score += 40; reasons.push(`평소 가격 대비 -${pct.toFixed(0)}%`); }
    else if (pct >= 35) { score += 20; reasons.push(`평소 가격 대비 -${pct.toFixed(0)}%`); }
  }
  if (o.price != null && o.crossBaseline && o.crossBaseline > 0) {
    const pct = ((o.crossBaseline - o.price) / o.crossBaseline) * 100;
    if (pct >= 60) { score += 60; reasons.push(`같은 모델 다른 판매처 대비 -${pct.toFixed(0)}%`); }
    else if (pct >= 40) { score += 40; reasons.push(`같은 모델 다른 판매처 대비 -${pct.toFixed(0)}%`); }
    else if (pct >= 25) { score += 20; reasons.push(`같은 모델 다른 판매처 대비 -${pct.toFixed(0)}%`); }
  }
  const official = isOfficialSeller(o.seller, o.site);
  if (official) score -= 20;
  else if (o.usualSeller && isOfficialSeller(o.usualSeller, o.site) && o.seller && o.seller !== o.usualSeller) { score += 30; reasons.push(`공식 판매자(${o.usualSeller})에서 제3자(${o.seller})로 바뀜`); }
  else if (o.seller && !NEUTRAL_SELLERS.has(o.seller)) { score += 10; reasons.push("공식 판매자 아님"); }
  const churn = o.sellerChanges ?? 0;
  if (!official && churn >= SELLER_CHURN_HIGH) { score += 25; reasons.push(`판매자가 자주 바뀜 (90일 ${churn}회)`); }
  else if (!official && churn >= SELLER_CHURN_MID) { score += 15; reasons.push(`판매자가 자주 바뀜 (90일 ${churn}회)`); }
  if (OVERSEAS.has(o.site)) { score += 15; reasons.push("해외 오픈마켓 (판매자 검증 약함)"); }
  if (o.modelNo && o.title) {
    const norm = (s: string) => s.toLowerCase().replace(/-/g, "").replace(/\s/g, "");
    if (!norm(o.title).includes(norm(o.modelNo))) { score += 10; reasons.push(`제목에 모델명 ${o.modelNo} 없음`); }
  }
  score = Math.max(0, score);
  const level: RiskLevel = score >= 60 ? "high" : score >= 30 ? "medium" : "low";
  if (!reasons.length) reasons.push("특이 신호 없음");
  return { level, score, reasons };
}

/** 최근 days 안에서 판매자가 바뀐 횟수 (연속 스냅샷의 판매자가 다르면 1회). 워커 baseline.seller_changes 와 동일. */
export function countSellerChanges(snaps: Snapshot[], days = 90, now = Date.now()): number {
  const cutoff = now - days * 86400_000;
  const sellers = snaps.filter((s) => s.seller && new Date(s.captured_at).getTime() >= cutoff).map((s) => s.seller as string);
  let n = 0;
  for (let i = 1; i < sellers.length; i++) if (sellers[i] !== sellers[i - 1]) n++;
  return n;
}

export function guessModelNo(title: string): string | null {
  const m = title.toUpperCase().match(/\b[A-Z]{1,4}-?\d{2,5}[A-Z]{0,3}\d{0,2}\b/);
  return m ? m[0] : null;
}

export const RISK_LABEL: Record<RiskLevel, string> = { low: "정품 리스크 낮음", medium: "정품 리스크 보통", high: "정품 리스크 높음" };
export const RISK_CLS: Record<RiskLevel, string> = {
  low: "bg-slate-100 text-slate-600 dark:bg-slate-800 dark:text-slate-300",
  medium: "bg-orange-100 text-orange-800 dark:bg-orange-900/50 dark:text-orange-200",
  high: "bg-rose-100 text-rose-800 dark:bg-rose-900/50 dark:text-rose-200",
};
